rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -145,90 +148,89 @@ class rmw_context_impl_s::Data final
 #ifndef _MSC_VER
-    if (strncmp(
-        z_string_data(z_loan(shm_enabled)),
-        "true",
-        z_string_len(z_loan(shm_enabled))) == 0)
-    {
-      // TODO(yuyuan): determine the default alignment of SHM
-      z_alloc_alignment_t alignment = {5};
-      z_owned_memory_layout_t layout;
-      z_memory_layout_new(&layout, SHM_BUFFER_SIZE_MB * 1024 * 1024, alignment);
-
-      z_owned_shm_provider_t provider;
-      if (z_posix_shm_provider_new(&provider, z_loan(layout)) != Z_OK) {
-        RMW_ZENOH_LOG_ERROR_NAMED("rmw_zenoh_cpp", "Unable to create an SHM provider.");
-        throw std::runtime_error("Unable to create an SHM provider.");
+    if (shm_enabled == "true") {
+      auto layout = zenoh::MemoryLayout(
+        SHM_BUFFER_SIZE_MB * 1024 * 1024,
+        zenoh::AllocAlignment({5}));
+      zenoh::PosixShmProvider provider(layout, &result);
+      if (result != Z_OK) {
+        throw std::runtime_error("Unable to create shm provider.");
       }
-      shm_provider_ = provider;
+      shm_provider_ = std::move(provider);
     }
-    auto free_shm_provider = rcpputils::make_scope_exit(
-      [this]() {
-        if (shm_provider_.has_value()) {
-          z_drop(z_move(shm_provider_.value()));
-        }
-      });
 #endif
     graph_guard_condition_ = std::make_unique<rmw_guard_condition_t>();
     graph_guard_condition_->implementation_identifier = rmw_zenoh_cpp::rmw_zenoh_identifier;
     graph_guard_condition_->data = &guard_condition_data_;
 
     // Setup the liveliness subscriber to receives updates from the ROS graph
     // and update the graph cache.
-    z_liveliness_subscriber_options_t sub_options;
-    z_liveliness_subscriber_options_default(&sub_options);
-    z_owned_closure_sample_t callback;
-    z_closure(&callback, graph_sub_data_handler, nullptr, this);
-    z_view_keyexpr_t liveliness_ke;
-    z_view_keyexpr_from_str(&liveliness_ke, liveliness_str.c_str());
-    if (z_liveliness_declare_subscriber(
-        session_->loan(),
-        &graph_subscriber_, z_loan(liveliness_ke),
-        z_move(callback), &sub_options) != Z_OK)
-    {
+    zenoh::KeyExpr keyexpr_cpp(liveliness_str.c_str());
+    zenoh::Session::LivelinessSubscriberOptions sub_options =
+      zenoh::Session::LivelinessSubscriberOptions::create_default();
+    sub_options.history = true;
+    graph_subscriber_cpp_ = session_->liveliness_declare_subscriber(
+      keyexpr_cpp,
+      [&](const zenoh::Sample & sample) {
+        // Look up the data shared_ptr in the global map. If it is in there, use it.
+        // If not, it is being shutdown so we can just ignore this update.
+        std::shared_ptr<rmw_context_impl_s::Data> data_shared_ptr{nullptr};
+        {
+          std::lock_guard<std::mutex> lk(data_to_data_shared_ptr_map_mutex);
+          if (data_to_data_shared_ptr_map.count(this) == 0) {
+            return;
+          }
+          data_shared_ptr = data_to_data_shared_ptr_map[this];
+        }
```

Comment:
This is one of those things that is going to be able to go away with lambda captures available.  But I think we should leave this as-is for now, and do a separate PR fixing it.